Split makeString() and friends out into their own header

weinig opened the pull request:
#### 632f10381f7b7a484ab345f3294b378b300d1c50
<pre>
Split makeString() and friends out into their own header
<a href="https://bugs.webkit.org/show_bug.cgi?id=275795">https://bugs.webkit.org/show_bug.cgi?id=275795</a>

Reviewed by Darin Adler.

Moved `makeString`, `tryMakeString`, `makeAtomString`, `tryMakeAtomString`
and `makeStringByInserting` into a new header, wtf/text/MakeString.h.

This is a prerequisite to supporting `interleave()` in makeString,
as this allows makeString to depend on StringBuilder without a cycle.

Also removes remaining users of the StringOperators style of concatenation
as there were only a few users remaining.

- Source list elided -

Canonical link: <a href="https://commits.webkit.org/280586@main">https://commits.webkit.org/280586@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5465edc81b259e7eb55941701196fd3cb6559d48

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57026 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36354 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9501 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60646 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7469 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59154 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43977 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7659 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46189 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5258 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59056 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34142 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49224 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27049 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30922 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6555 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6474 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/50118 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52884 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6826 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62327 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/56267 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/939 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6930 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53448 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/942 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49279 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53494 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12610 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/801 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/78028 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32183 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12929 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33268 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34353 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33014 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -3810,8 +3811,7 @@ static void runInteractive(GlobalObject* globalObject)
                 break;
-            source = source + String::fromUTF8(line);
-            source = source + '\n';
+            source = makeString(source, String::fromUTF8(line), '\n');
```
This can be done more efficiently by using something like `byteCast<uchar8_t>(span(line))` instead, which will convert as part of the concatenation process without allocating a temporary string.

darinadler commented on Source/WTF/WTF.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 55;
+	objectVersion = 54;
```
Maybe don't land this change.

darinadler commented on Source/WTF/wtf/text/MakeString.h:
```diff
@@ -0,0 +1,143 @@
+    ASSERT(length <= String::MaxLength);
+    if (areAllAdapters8Bit) {
+        LChar* buffer;
+        RefPtr resultImpl = StringImpl::tryCreateUninitialized(length, buffer);
```
Why not just name this `result`?

darinadler commented on Source/WTF/wtf/text/MakeString.h:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    UChar* buffer;
+    RefPtr resultImpl = StringImpl::tryCreateUninitialized(length, buffer);
```
Why not just name this `result`?

darinadler commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -659,7 +660,7 @@ String CSSSelector::selectorText(StringView separator, StringView rightSide) con
     } else if (auto separatorText = separatorTextForNestingRelative(); !separatorText.isNull()) {
         // We have a separator but no tag history which can happen with implicit relative nesting selector
-        return separatorText + builder.toString();
+        return makeString(separatorText, builder.toString());
```
Should we teach `makeString` to work directly with a `StringBuilder` argument without creating an additional intermediate `String`? Should we add `StringBuilder::prepend` and can it be efficient enough to be OK?

darinadler commented on Source/WebCore/dom/QualifiedName.h:
```diff
@@ -160,7 +161,7 @@ inline String QualifiedName::toString() const
         return localName();
 
-    return prefix().string() + ':' + localName().string();
+    return makeString(prefix().string(), ':', localName().string());
```
Can we omit the calls to `string()`?

darinadler commented on Source/WebCore/fileapi/FileCocoa.mm:
```diff
@@ -30,6 +30,7 @@
 
 #import <UniformTypeIdentifiers/UniformTypeIdentifiers.h>
 #import <wtf/FileSystem.h>
+#include <wtf/text/MakeString.h>
```
Should use `#import`.

darinadler commented on Source/WebCore/loader/TextResourceDecoder.cpp:
```diff
@@ -662,8 +663,8 @@ String TextResourceDecoder::flush()
-    String decoded = decode(data);
-    return decoded + flush();
+    auto decoded = decode(data);
+    return makeString(decoded, flush());
```
Does `makeString` optimize the case where one argument is a `String` and the other arguments have zero characters? If not, perhaps it should since `flush()` almost always returns the empty string. Or could add a fast path here for the same reason. Is it possible that the `+` operator had that case optimized?

darinadler commented on Source/WebCore/platform/cocoa/KeyEventCocoa.mm:
```diff
@@ -32,6 +32,7 @@
 #import <wtf/ASCIICType.h>
 #import <wtf/HexNumber.h>
 #import <wtf/MainThread.h>
+#include <wtf/text/MakeString.h>
```
Should use `#import`.

darinadler commented on Source/WebCore/platform/xdg/MIMETypeRegistryXdg.cpp:
```diff
@@ -38,7 +40,7 @@ String MIMETypeRegistry::mimeTypeForExtension(StringView string)
 
     // Build any filename with the given extension.
-    String filename = makeString("a."_s, string);
+    auto filename = makeString("a."_s, string);
```
Some day maybe we will want `makeUTF8String` for cases like this one.

darinadler commented on Source/WebCore/svg/SVGAnimateTransformElement.cpp:
```diff
@@ -67,7 +67,7 @@ void SVGAnimateTransformElement::attributeChanged(const QualifiedName& name, con
 String SVGAnimateTransformElement::animateRangeString(const String& string) const
 {
-    return SVGTransformValue::prefixForTransformType(m_type) + string + ')';
+    return makeString(SVGTransformValue::prefixForTransformType(m_type), string, ')');
```
Clearly this function should take a `StringView`.

darinadler commented on Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
```diff
@@ -135,7 +136,7 @@ String CacheStorageDiskStore::recordFilePath(const NetworkCache::Key& key) const
 String CacheStorageDiskStore::recordBlobFilePath(const String& recordPath) const
 {
-    return recordPath + blobSuffix;
+    return makeString(recordPath, blobSuffix);
```
Clearly this function should take a `StringView`.

darinadler commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -434,7 +434,7 @@ String Storage::recordPathForKey(const Key& key) const
 static String blobPathForRecordPath(const String& recordPath)
 {
-    return recordPath + blobSuffix;
+    return makeString(recordPath, blobSuffix);
```
Clearly this function should take a `StringView`.

darinadler commented on Source/WebKit/UIProcess/Automation/win/WebAutomationSessionWin.cpp:
```diff
@@ -349,7 +350,7 @@ OptionSet<WebEventModifier> WebAutomationSession::platformWebModifiersFromRaw(We
 {
     // https://www.w3.org/TR/uievents-code/#key-alphanumeric-writing-system
-    String keyIdentifier = "Key"_s + keySequence.convertToASCIIUppercase();
+    auto keyIdentifier = makeString("Key"_s, keySequence.convertToASCIIUppercase());
```
This should use `uppercase()` to avoid creating an intermediate `StringImpl`. Or that might not be the exact name.

darinadler commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -907,7 +908,7 @@ static NavigatingToAppBoundDomain schemeOrDomainIsAppBound(const String& host, c
     if (path.isEmpty())
         path = NSHomeDirectory();
```
It’s possible we can use `makeString` directly on the `NSString *` in this case for better efficiency.

darinadler reviewed (approved):
Not all the comments are urgent to address in this pull request.

weinig committed:
Split makeString() and friends out into their own header
https://bugs.webkit.org/show_bug.cgi?id=275795

Reviewed by Darin Adler.

Moved `makeString`, `tryMakeString`, `makeAtomString`, `tryMakeAtomString`
and `makeStringByInserting` into a new header, wtf/text/MakeString.h.

This is a prerequisite to supporting `interleave()` in makeString,
as this allows makeString to depend on StringBuilder without a cycle.

Also removes remaining users of the StringOperators style of concatenation
as there were only a few users remaining.

- Source list elided -

Canonical link: https://commits.webkit.org/280586@main
